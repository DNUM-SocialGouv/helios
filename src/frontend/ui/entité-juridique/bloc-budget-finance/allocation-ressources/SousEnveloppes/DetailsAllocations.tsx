import { ReactElement, useState } from "react";

import { IEnveloppe } from "../../../../../../backend/métier/entities/entité-juridique/EntitéJuridiqueAllocationRessources";
import { useDependencies } from "../../../../commun/contexts/useDependencies";
import {
  HistogrammeHorizontalRowMultiple,
  colorsAllocations,
} from "../../../../indicateur-métier/qualite/ReclamationsParAnnee/HistogrammeHorizontalRowMultiple/HistogrammeHorizontalRowMultiple";
import { convertFloatToComma, formatNumbuerWithSpaces } from "../../AllocationRessourcesViewModel";
import styles from "./DetailsAllocations.module.css";

type ShowDetailsTitleProps = Readonly<{ for: string; children: ReactElement; direction?: string }>;
const ShowDetailsTitle = ({ for: identifiant, children, direction = "right" }: ShowDetailsTitleProps) => {
  const [isOpen, setIsOpen] = useState(false);

  return (
<span
  aria-controls={identifiant}
  aria-expanded={isOpen}
  className={
    styles["titleDetails"] +
    " fr-btn-NO fr-btn--sm-NO fr-btn--tertiary-no-outline-NO fr-btn--icon-" +
    direction +
    (isOpen ? " fr-icon-arrow-up-s-line" : " fr-icon-arrow-down-s-line")
  }
  data-fr-opened={isOpen}
  data-testid="show-details-title"
  icon-direction={direction}
  onClick={() => {
    setIsOpen(!isOpen);
  }} 
  onKeyDown={() => {}}
  role="button"
  tabIndex={0} 
>
  {children}
</span>

  );
};

type ShowDetailsProps = Readonly<{ dataTitle: ReactElement; children: any; id: string; className: string; dataName?: string; directionIcone?: string }>;
const ShowDetails = ({ dataTitle, children, id, className, dataName, directionIcone = "right" }: ShowDetailsProps) => {
  return (
    <div className={className} data-testid="show-details">
      <div className={styles["envTitle"]}>{dataName}</div>
      <div className={styles["envContent"]}>
        <ShowDetailsTitle direction={directionIcone} for={id}>
          {dataTitle}
        </ShowDetailsTitle>
        <ul className="fr-collapse niveau3" id={id}>
          {children}
        </ul>
      </div>
    </div>
  );
};
 
type DetailsAllocationsProps = Readonly<{
  data: IEnveloppe[];
}>;

export function DetailsAllocations({ data }: DetailsAllocationsProps) {
  const { wording } = useDependencies();
 
  return (
    <div className={styles['bigContainer']}>
      <div className={styles['repBigTitle']}>{wording.REPARTITION_DES_SOUS_ENVELOPPES}</div>

      {data.map((enveloppe) => (
        <ShowDetails
          className={styles["envContainer"]}
          dataName={`Enveloppe ${enveloppe.enveloppe}`}
          dataTitle={
            <div className={styles["titleContainer"]}>
              <div className={styles["titleDetails"]}>
                <HistogrammeHorizontalRowMultiple
                  data={enveloppe.sousEnveloppes.map((sousEnveloppe) => ({ key: sousEnveloppe.sousEnveloppe, value: sousEnveloppe.pourcentage }))}
                  realPercentage={enveloppe.pourcentage}
                />
               <span className={styles["envTotal"]}>{formatNumbuerWithSpaces(enveloppe.total)} €</span>
          
              </div>
       
            </div>
          }
          directionIcone="left"
          id={enveloppe.enveloppe}
          key={enveloppe.enveloppe}
        >
          {enveloppe.sousEnveloppes.map((sousEnveloppe, index) => (
            <ShowDetails
              className={styles["subEnvContainer"]}
              dataTitle={
                <>
                  <div className={styles["carreSousEnveloppe"]} style={{ backgroundColor: colorsAllocations[index], color: colorsAllocations[index] }}>
                  i
                  </div>
                  <span className={styles["subEnvTitle"]}>
                    Sous enveloppe {sousEnveloppe.sousEnveloppe}{" "}
                    <span className={styles["totalSousEnveloppe"]}>
                      {" "}
                      {formatNumbuerWithSpaces(sousEnveloppe.total)} € ({convertFloatToComma(sousEnveloppe.pourcentage)}%){" "}
                    </span>{" "}
                  </span>
                </>
              }
              id={enveloppe.enveloppe + sousEnveloppe.sousEnveloppe}
              key={sousEnveloppe.sousEnveloppe}
            >
              <div className={styles["modesDeDelegationContainer"]}>
                {sousEnveloppe.modesDeDélégation.map((mode) => (
                  <div className={styles["modesDeDelegationItem"]} key={mode.modeDeDélégation}>
                    {mode.modeDeDélégation} : {formatNumbuerWithSpaces(mode.montantNotifié)} € ({convertFloatToComma(mode.pourcentage)}%)
                  </div>
                ))}
        
              </div>
            </ShowDetails>
          ))}
        </ShowDetails>
      ))}
    </div>
  );
}
