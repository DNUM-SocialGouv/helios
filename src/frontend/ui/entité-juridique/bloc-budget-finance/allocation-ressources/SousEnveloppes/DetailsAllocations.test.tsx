import { render, screen } from '@testing-library/react';
import React from 'react';

import { DetailsAllocations } from './DetailsAllocations';

// Mocking the useDependencies hook
jest.mock('../../../../commun/contexts/useDependencies', () => ({
  useDependencies: () => ({
    wording: {
      REPARTITION_DES_SOUS_ENVELOPPES: 'Mocked Repartition des sous enveloppes',
    },
  }),
}));

const testData = [
  {
    enveloppe: 'Enveloppe 1',
    pourcentage: 100,
    total: 1000,
    sousEnveloppes: [
      {
        sousEnveloppe: 'Sous Enveloppe 1',
        pourcentage: 100,
        total: 1000,
        modesDeDélégation: [
          { modeDeDélégation: 'Mode 1', montantNotifié: 500, pourcentage: 50 },
          { modeDeDélégation: 'Mode 2', montantNotifié: 500, pourcentage: 50 },
        ],
      },
    ],
  },
];

const testDataMultiple = [
  {
    enveloppe: 'Enveloppe 1',
    pourcentage: 66.66,
    total: 1000,
    sousEnveloppes: [
      {
        sousEnveloppe: 'Sous Enveloppe 1',
        pourcentage: 100,
        total: 1000,
        modesDeDélégation: [
          { modeDeDélégation: 'Mode 1', montantNotifié: 700, pourcentage: 70 },
          { modeDeDélégation: 'Mode 2', montantNotifié: 300, pourcentage: 30 },
        ],
      },
    ],
  },
  {
    enveloppe: 'Enveloppe 2',
    pourcentage: 33.33,
    total: 500,
    sousEnveloppes: [
      {
        sousEnveloppe: 'Sous Enveloppe 2',
        pourcentage: 100,
        total: 500,
        modesDeDélégation: [
          { modeDeDélégation: 'Mode 3', montantNotifié: 500, pourcentage: 100 },
        ],
      },
    ],
  },
];

const testDataEdgeCases = [
  {
    enveloppe: 'Enveloppe 1',
    pourcentage: 10,
    total: 10,
    sousEnveloppes: [
      {
        sousEnveloppe: 'Sous Enveloppe 1',
        pourcentage: 100,
        total: 100,
        modesDeDélégation: [],
      },
    ],
  },
  {
    enveloppe: 'Enveloppe 2',
    pourcentage: 10,
    total: 10,
    sousEnveloppes: [
      {
        sousEnveloppe: 'Sous Enveloppe 2',
        pourcentage: 10,
        total: 100,
        modesDeDélégation: [],
      },
    ],
  },
  {
    enveloppe: 'Enveloppe 3',
    pourcentage: 80,
    total: 800,
    sousEnveloppes: [
      {
        sousEnveloppe: 'Sous Enveloppe 3',
        pourcentage: 100,
        total: 800,
        modesDeDélégation: [],
      },
    ],
  },
];

describe('DetailsAllocations component', () => {
  it('renders correctly with provided data', async () => {
    render(<DetailsAllocations data={testData} />);
    
    // Attendre que le titre principal soit rendu
    await screen.findByText('Mocked Repartition des sous enveloppes');
    
    // Vérifier que le nom de l'enveloppe est rendu
    expect(screen.getByText('Enveloppe Enveloppe 1')).toBeInTheDocument();
    
    // Vérifier que le nom de la sous-enveloppe est rendu
    expect(screen.getByText('Sous enveloppe Sous Enveloppe 1')).toBeInTheDocument();
    
  });

  it('renders correctly with multiple envelopes', async () => {
    render(<DetailsAllocations data={testDataMultiple} />);
    
    // Attendre que les titres des enveloppes soient rendus
    await screen.findByText('Enveloppe Enveloppe 1');
    expect(screen.getByText('Enveloppe Enveloppe 2')).toBeInTheDocument();
    
    // Vérifier que les titres et les détails des sous-enveloppes sont rendus pour chaque enveloppe
    expect(screen.getByText('Sous enveloppe Sous Enveloppe 1')).toBeInTheDocument();
    expect(screen.getByText('Sous enveloppe Sous Enveloppe 2')).toBeInTheDocument();
  });

  it('renders correctly with different percentage thresholds', async () => {
    render(<DetailsAllocations data={testDataEdgeCases} />);
    expect(screen.getByText('Enveloppe Enveloppe 1')).toBeInTheDocument();
    expect(screen.getByText('Enveloppe Enveloppe 2')).toBeInTheDocument();
    expect(screen.getByText('Enveloppe Enveloppe 3')).toBeInTheDocument();
    expect(screen.getByText('Sous enveloppe Sous Enveloppe 1')).toBeInTheDocument();
    expect(screen.getByText('Sous enveloppe Sous Enveloppe 2')).toBeInTheDocument();
    expect(screen.getByText('Sous enveloppe Sous Enveloppe 3')).toBeInTheDocument();
 
  });
});
