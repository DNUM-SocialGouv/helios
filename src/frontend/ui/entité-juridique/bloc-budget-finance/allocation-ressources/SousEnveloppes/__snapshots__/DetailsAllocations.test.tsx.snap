// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsAllocations component matches snapshot 1`] = `
<div>
  <div
    class="bigContainer"
  >
    <div
      class="repBigTitle"
    >
      Mocked Repartition des sous enveloppes
    </div>
    <div
      class="envContainer"
      data-testid="show-details"
    >
      <div
        class="envTitle"
      >
        Enveloppe Enveloppe 1
      </div>
      <div
        class="envContent"
      >
        <span
          aria-controls="Enveloppe 1"
          aria-expanded="false"
          class="titleDetails fr-btn-NO fr-btn--sm-NO fr-btn--tertiary-no-outline-NO fr-btn--icon-left fr-icon-arrow-down-s-line"
          data-fr-opened="false"
          data-testid="show-details-title"
          icon-direction="left"
          role="button"
          tabindex="0"
        >
          <div
            class="titleContainer"
          >
            <div
              class="titleDetails"
            >
              <div
                class="histogramme_horizontal_row_container"
                style="width: 80%;"
              >
                <ul>
                  <li>
                    <div
                      class="barHistogramme"
                      data-testid="bar-Sous Enveloppe 1"
                      style="width: 50%; background-color: rgb(153, 179, 249);"
                    >
                      i
                    </div>
                  </li>
                </ul>
              </div>
            </div>
            <span
              class="envTotal2"
            >
              1 000
               €
            </span>
          </div>
        </span>
        <ul
          class="fr-collapse niveau3"
          id="Enveloppe 1"
        >
          <div
            class="subEnvContainer"
            data-testid="show-details"
          >
            <div
              class="envTitle"
            />
            <div
              class="envContent"
            >
              <span
                aria-controls="Enveloppe 1Sous Enveloppe 1"
                aria-expanded="false"
                class="titleDetails fr-btn-NO fr-btn--sm-NO fr-btn--tertiary-no-outline-NO fr-btn--icon-right fr-icon-arrow-down-s-line"
                data-fr-opened="false"
                data-testid="show-details-title"
                icon-direction="right"
                role="button"
                tabindex="0"
              >
                <div
                  class="carreSousEnveloppe"
                  style="background-color: rgb(153, 179, 249); color: rgb(153, 179, 249);"
                >
                  i
                </div>
                <span
                  class="subEnvTitle"
                >
                  Sous enveloppe 
                  Sous Enveloppe 1
                   
                  <span
                    class="totalSousEnveloppe"
                  >
                     
                    500
                     € (
                    50
                    %)
                     
                  </span>
                   
                </span>
              </span>
              <ul
                class="fr-collapse niveau3"
                id="Enveloppe 1Sous Enveloppe 1"
              >
                <div
                  class="modesDeDelegationContainer"
                >
                  <div
                    class="modesDeDelegationItem"
                  >
                    Mode 1
                     : 
                    200
                     € (
                    20
                    %)
                  </div>
                  <div
                    class="modesDeDelegationItem"
                  >
                    Mode 2
                     : 
                    300
                     € (
                    30
                    %)
                  </div>
                </div>
              </ul>
            </div>
          </div>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
